### PR TITLE
feat(encoding): support BOM detection

### PR DESCRIPTION
**Pull request by WingLim**
### What does this PR do?

Support BOM detection and `TextEncoder.ignoreBOM`.

Closes https://github.com/oven-sh/bun/issues/5471

- [x] Code changes

### How did you verify your code works?

- [x] Add test

### CODE REVIEW

File: test/js/web/encoding/text-decoder.test.js [hunk 1]
```diff
@@ -250,7 +250,7 @@ describe("TextDecoder", () => {
   it("constructor should set values", () => {
     const decoder = new TextDecoder("utf-8", { fatal: true, ignoreBOM: false });
     expect(decoder.fatal).toBe(true);
-    // expect(decoder.ignoreBOM).toBe(false); // currently the getter for ignoreBOM doesn't work and always returns undefined
+    expect(decoder.ignoreBOM).toBe(false);
```
Comment (Jarred-Sumner) [inline test/js/web/encoding/text-decoder.test.js:253]: we haven't actually implemented BOM detection

Comment (WingLim) [reply test/js/web/encoding/text-decoder.test.js:253]: Oh, okay, let me try to implement the BOM detection in this PR, I will convert this to draft for now.

File: src/bun.js/webcore/encoding.zig [hunk 2]
```diff
@@ -559,6 +559,13 @@ pub const TextDecoder = struct {
                     remainder = remainder[1..];
                     continue;
                 },
+                // BOM handling
+                0xFEFF => {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/encoding.zig:563]: This should only be in there for the first character right? This code runs on every character

Comment (WingLim) [reply src/bun.js/webcore/encoding.zig:563]: We handle the first character of BOM in `decodeSlice`.

In unicode `0xFEFF` is a special, we should not rewrite it to `0xFFFD` in else block, just keep it. This is to be consistent with node.

**Comment by Jarred-Sumner** [comment]
Thank you

**Pull request merged**